fms/modules/linear.py

```diff
@@ -0,0 +1,178 @@
+
+    return linear_type.lower()
+
+
+def get_linear(
+    in_features: int,
+    out_features: int,
+    bias: bool,
+    linear_config:  Optional[Mapping[str, Any]] = None,
+) -> nn.Module:
+    linear_type = get_linear_type(linear_config)
+
+    # TODO: how to merge these calls that get different arguments?
+    if linear_type in __type_factory_map:
+        if linear_type == "torch_linear":
+            return __type_factory_map[linear_type](in_features, out_features, bias)
+        else:
+            return __type_factory_map[linear_type](
+                in_features, out_features, bias, linear_config
+            )
+    else:
+        raise KeyError(f"Unsupported linear type `{linear_type}`")
+
+
+@dataclass
+class LinearModuleShardingInfo:
+    linear_module: torch.nn.Module
+    sharding_dim: int
+    max_partitions: List[int]
+
+
+@dataclass
+class LinearParameterShardingInfo:
+    sharding_dim: int
+    shard_type: ShardType
+
+
+def shard_base_linear(
+    tensor_values: Dict[str, torch.Tensor],
+    tp_module: TPModule,
+    module_sharding_info: Dict[str, LinearModuleShardingInfo],
+    param_sharding_info: Dict[str, Dict[str, LinearParameterShardingInfo]],
+) -> None:
+    all_params = {}
+    used_keys: set[str] = set()
+
+    # Collect quantization parameters to copy on sharded module
+    tensor_device = None
+    param_count = 0
+    for module_name in module_sharding_info:
+        for param_name in param_sharding_info[module_name]:
+            if module_name not in all_params:
+                all_params[module_name] = {}
+            # TODO: reusing method '_get_sd_weight' but consider changing its name
+            all_params[module_name][param_name] = tp_module._get_sd_weight(
+                tensor_values, used_keys, [module_name, param_name]
+            )
+            # TODO: We should transfer the initial_device here from get_model
+            if tensor_device is None:
+                tensor_device = all_params[module_name][param_name].device
+            param_count += 1
+
+    # FIXME: gptq llama ckpt may have bias but llama linear layers do not
+    # We can't raise an error here, because it stops further computations
+    # Better handling is needed (drop bias tensor earlier, if zeros?)
+    if len(tensor_values) > param_count:
+        unused_keys = set(tensor_values.keys()).difference(used_keys)
+        # print("UNUSED KEYS: ", ', '.join(unused_keys))
+        # raise AttributeError(f"Unused weight(s): {', '.join(unused_keys)}")
+
+    # Shard module, one parameter at the time
+    for module_name, module_info in module_sharding_info.items():
+        for param_name, param_info in param_sharding_info[module_name].items():
+            module_param = getattr(module_info.linear_module, param_name)
+            if module_param.device == torch.device("meta"):
+                # TODO: We should bring the default dtype if set from get_model here
```

Comment:
why is this logic still here? I superseded that with lines 512-515 and 521 in serialization.py. All parameters should already be out of meta at this point.